### PR TITLE
[AIR] Delegate Preprocessor conversions to `transform_batch`

### PR DESCRIPTION
**Pull request by amogkam**
Signed-off-by: Amog Kamsetty <amogkamsetty@yahoo.com>

Delegate all conversion logic in preprocessor to `transform_batch`. The base `transform` can just map over the native dataset format. This keeps all conversions isolated to one place in the code.

Also adds consistency in Arrow->Numpy transform in both Datasets and Preprocessors

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by jiaodong** [comment]
we have some hardcoded string issue in dataset and that's why we shouldn't use hardcoded string for enum value like this :/

```
ValueError: The batch format must be one of ['default', 'native', 'pandas', 'pyarrow', 'numpy'], got: arrow
```

File: python/ray/data/preprocessor.py
Comment (jiaodong) [inline python/ray/data/preprocessor.py:248]: dataset has `arrow` format, but we don't have equivalent transform type / batch_format right ? Basically all data formats in this error message from CI  

```
ValueError: The batch format must be one of ['default', 'native', 'pandas', 'pyarrow', 'numpy'], got: arrow
```

needs to map to either `pandas` or `numpy` batch_format

Comment (amogkam) [reply python/ray/data/preprocessor.py:248]: ah yeah dataset format is arrow, but batch format should be pyarrow...

Comment (jiaodong) [reply python/ray/data/preprocessor.py]: > 

should we ensure these two are just the same ? I don't see a reason why a python codebase that don't use pyarrow to handle arrow format ..

File: python/ray/data/datasource/datasource.py
Comment (clarkzinzow) [inline python/ray/data/datasource/datasource.py:251]: This isn't a tensor array FYI, this is what you'd get with e.g. `ray.data.range_table()`, and we ensure that the column name is human readable (like `"value"`), not our internal `TENSOR_COL_NAME`. Could you revert this back?

File: python/ray/data/_internal/arrow_block.py
Comment (clarkzinzow) [inline python/ray/data/_internal/arrow_block.py:211]: So this will break the current Arrow --> NumPy conversion semantics when `columns = None`, which is currently:
- Return a single NumPy ndarray for a single-column table.
- Return a dict of NumPy ndarray for a multi-column table.

This PR would change it to always return a dict of NumPy ndarray, even when the column name is our internal `TENSOR_COL_NAME`, which should definitely be returned as a single NumPy ndarray.

IMO we should align to the `_convert_batch_type_to_numpy()` semantics when `columns: List[str]` and `columns = None`, which are:
- Return a single NumPy ndarray for a single-column table with column name `TENSOR_COL_NAME`.
- Return a dict of NumPy ndarray otherwise.

This PR also currently changes the `columns: str` case, which should always return a single NumPy ndarray but with this PR would return a dict of ndarrays.

Overall, I think that the semantics should be:
- `columns = None or columns = List[str]`: `_convert_batch_type_to_numpy()` semantics.
- `columns = str`: Always return a single NumPy ndarray.

```python
        if columns is None:
            columns = self._table.column_names
        # Should add this helper to ArrowBlock.
        self._check_has_columns(columns)
        if not isinstance(columns, list):
            batch = self._table.column(columns)
        else:
            batch = self._table.select(columns)
        return _convert_batch_type_to_numpy(batch)
```

Also, we should make this same change in semantics + use of `_convert_batch_type_to_numpy()` in [`PandasBlock.to_numpy()`](https://github.com/ray-project/ray/blob/d527ce8b7d6367f83292e5da02fc3ebe6f11860a/python/ray/data/_internal/pandas_block.py#L193-L213).

File: python/ray/air/util/data_batch_conversion.py
Comment (clarkzinzow) [inline python/ray/air/util/data_batch_conversion.py:149]: I'm about to merge a fix to this part of `ArrowBlock.to_numpy()` btw, I found that `pyarrow.array([], type=col.type)` does not work for the `col.num_chunks == 0` case when the column is an extension array, so we need to reorder these and have `_concatenate_extension_column` properly handle the `col.num_chunks == 0` for the extension array case. Relevant bits of the PR about to merged are here:
- https://github.com/ray-project/ray/pull/29055/files#diff-a637aa9db5408fe440f8b335a480e66e6636d97e71dc4bad164c356ac43deafdR211-R214
- https://github.com/ray-project/ray/pull/29055/files#diff-bf26e743f5dc78351cf2755d92da8f54a0e1d37baa8b809dc679e64ddce83840

You could port these to your PR or wait until I merged (in the next hour or two) and then rebase.

**Review by clarkzinzow — CHANGES_REQUESTED** [review]
Really great simplification and consolidation! Once we have the `"zero_copy"` batch format, porting to that should end up being very simple.

Got a few changes that I think we need to make, but overall looking good.

File: python/ray/data/preprocessor.py
Comment (jiaodong) [inline python/ray/data/preprocessor.py:248]: do we still need this function if it's caller is removed in `preprocessor.py` ? This is the only place we call into it within this file.

In addition, the other use of it is `BatchMapper` that we provide user the ability to decide a batch_format that overrides our default dispatch that should also be checked

Comment (amogkam) [reply python/ray/data/preprocessor.py:248]: its still being called in `_transform_batch` right

**Comment by amogkam** [comment]
@clarkzinzow all the points make sense, but right now there is a inconsistency in how we do Arrow->numpy conversions in Datasets vs. Preprocessors.

Can we pick one approach and go with it?

**Pull request closed without merging**
